### DISCUSSION TITLE
Add support for suppressing console.log statements on client-side only

### DISCUSSION THREAD

**Discussion by kian-supportinglines** [vercel/next.js#48410] [Ideas]
### Goals

1. Add support in next config file to be able to suppress console.log statements in client and server side separately
2.
3.

### Non-Goals

1.
2.
3.

### Background

When creating a production app, typically we want to suppress console.log statements, but want to keep them for server-side for easier debugging when something goes wrong. The only alternative is to manually remove client-side console.log statements in production.

### Proposal

I would love to contribute if I can know where to look to implement the changes. For instance, how does console.log suppression work?

**Comment by luuap** [comment]
There is a compiler option to [remove console.log calls](https://nextjs.org/docs/advanced-features/compiler#remove-console). It is available starting from v12.1.0.

**Reply by kian-supportinglines** [reply]
This removes all console.log calls. As stated in the discussion, I would like to remove these calls in server-side code only (i.e. in getServerSideProps functions)

**Reply by luuap** [reply]
Oh I see! I have a couple of suggestions:

For dev/prod logic separation you can use [environment variables](https://nextjs.org/docs/basic-features/environment-variables):
```

if (process.env.NODE_ENV === 'development') {
  // Only runs is development
}

if (process.env.NODE_ENV === 'production') {
  // Only runs is production
}

```

And for server/client separation, you can do a check for variables specific to that environment (i.e. browser, node):
```

if (typeof process === 'object') {
  // Only runs in node
}

if (typeof window === 'object') {
  // Only runs in browser
}

```

I hope this helps 😄

**Reply by kian-supportinglines** [reply]
Yes this does work. However, this just gave me an idea. Because Next uses a plain old js file for their config, we could do something like:

```
module.exports = {
  compiler: {
    removeConsole: (process.env.NODE_ENV === 'development') ? true : false ,
  },
}
```

This will remove all console.* statements in the dev environment. I've tested this before on Next 12, should work fine on Next 13 as well. If you just want to keep console.error calls, replace `true` with `{exclude: ['error']}`. 

Now, we can replace the condition with `typeof process === 'object'`. I will test this tomorrow, but I am sure it will work. Thanks!

**Reply by luuap** [reply]
Branching off your idea of doing that at the config level, you can also make your config a function. This function has a `phase` parameter you can use. Read more [here](https://nextjs.org/docs/api-reference/next.config.js/introduction).

**Reply by Manitej66** [reply]
Hey @luuap , could you send the config you used to disable client-side only logs?
